### ISSUE TITLE
torch.compile(torch.func.grad(fn)) hangs indefinitely on models with pooling ops — sympy.simplify exponential recursion in stride_at

### ISSUE THREAD

**Issue by jasminetrail** [pytorch/pytorch#181653]
### 🐛 Describe the bug

torch.compile(torch.func.grad(fn)) hangs indefinitely during inductor's C++ code generation phase when fn contains pooling layers (MaxPool2d, AvgPool2d). The hang occurs inside stride_at() in torch/_inductor/codegen/cpp.py, which calls sympy.simplify(new_index - index). The symbolic stride expression produced by the VJP transform of pooling ops over non-power-of-2 input dimensions causes SymPy's simplify → replace → walk tree walker to recurse exponentially and never return.

Oracles 1–4 and standalone torch.func.grad all complete correctly. The hang is specific to torch.compile(torch.func.grad(fn)).

### Call stack (truncated)

torch.compile(torch.func.grad(fn))(t)
  → torch/_inductor/compile_fx.py: fx_codegen_and_compile
  → torch/_inductor/graph.py: codegen
  → torch/_inductor/scheduler.py: _codegen_partitions
  → torch/_inductor/codegen/cpp.py: codegen_node
  → cpp.py:select_tiling → _try_get_stride → stride_at_vec_range → stride_at
  → sympy.simplify(new_index - index)         # <-- enters infinite recursion
  → sympy/core/basic.py: walk → walk → walk → ...
  → sympy/simplify/simplify.py: simplify → replace → walk → ...
  [exponential recursion, never returns]

### Reproducers (all confirmed to hang; no crash on oracles 1–4 or func.grad alone)

```python
# Minimal (crash_069d43ec)
import torch, torch.nn as nn, torch.func as tf

m1, m2, m3 = nn.MaxPool2d(2), nn.AvgPool2d(2), nn.MaxPool2d(2)
t = torch.randn([8, 10, 11, 10])

def fn(x): return m3(m2(m1(x))).mean()

torch.func.grad(fn)(t)                  # OK
torch.compile(torch.func.grad(fn))(t)   # hangs

# Variant 1 (crash_8161ebbd): tanh → RReLU → AvgPool2d → AvgPool2d → MaxPool2d
import torch, torch.nn as nn, torch.func as tf

m1, m2, m3, m4 = nn.RReLU().eval(), nn.AvgPool2d(2), nn.AvgPool2d(2), nn.MaxPool2d(2)
t = torch.randn([10, 12, 16, 12])

def fn(x): return m4(m3(m2(m1(torch.tanh(x))))).mean()
torch.compile(torch.func.grad(fn))(t)   # hangs

# Variant 2 (crash_a22e8c5b): AvgPool2d × 3 → sqrt(clamp)
import torch, torch.nn as nn, torch.func as tf

m1, m2, m3 = nn.AvgPool2d(2), nn.AvgPool2d(2), nn.AvgPool2d(2)
t = torch.randn([5, 12, 15, 10])

def fn(x): return torch.sqrt(torch.clamp(m3(m2(m1(x))), min=1e-6)).mean()

torch.compile(torch.func.grad(fn))(t)   # hangs
```

### Versions

Collecting environment information...
PyTorch version: 2.11.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: 15.0.0 ([git@github.com](mailto:git@github.com):llvm/llvm-project.git 4ba6a9c9f65bbc8bd06e3652cb20fd4dfc846137)
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Mar 3 2026, 11:56:32) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-94-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 52 bits physical, 57 bits virtual
Byte Order: Little Endian
CPU(s): 384
On-line CPU(s) list: 0-383
Vendor ID: AuthenticAMD
Model name: AMD EPYC 9684X 96-Core Processor
CPU family: 25
Model: 17
Thread(s) per core: 2
Core(s) per socket: 96
Socket(s): 2
Stepping: 2
BogoMIPS: 5099.98
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc amd_ibpb_ret arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d debug_swap ibpb_exit_to_user
Virtualization: AMD-V
L1d cache: 6 MiB (192 instances)
L1i cache: 6 MiB (192 instances)
L2 cache: 192 MiB (192 instances)
L3 cache: 2.3 GiB (24 instances)
NUMA node(s): 2
NUMA node0 CPU(s): 0-95,192-287
NUMA node1 CPU(s): 96-191,288-383
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit: Not affected
Vulnerability L1tf: Not affected
Vulnerability Mds: Not affected
Vulnerability Meltdown: Not affected
Vulnerability Mmio stale data: Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Mitigation; Safe RET
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected
Vulnerability Vmscape: Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.19.0.56
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.0
[pip3] nvidia-nccl-cu13==2.28.9
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.11.0
[pip3] triton==3.6.0
[conda] Could not collect

cc @mikaylagawarecki @chauhang @penguinwu @Chillee @samdow @kshitij12345 @ezyang @bobrenjc93 @aditvenk @laithsakka @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by jasminetrail** [comment]
Here is the [gist](https://colab.research.google.com/gist/jasminetrail/87781c6ae3f378409bbc6a3d9d07ec5d/issue_181653.ipynb)

**Comment by jansel** [comment]
My agent was unable to reproduce this on f513851aa85d9d3336376b54d3a431168cadf3ba. Reopen if you see this on the latest PyTorch nightly

**Comment by jasminetrail** [comment]
@jansel This issue still reproduces on latest nightly. Here is the [gist](https://colab.research.google.com/gist/jasminetrail/57f18ac7802fa1061da29d1b0e2e2533/issue-181653.ipynb)
